File: src/deejayd/sources/_medialist.py
# Deejayd, a media player daemon
# Copyright (C) 2007-2008 Mickael Royer <mickael.royer@gmail.com>
#                         Alexandre Rossi <alexandre.rossi@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import random

class SimpleMediaList(object):

    def __init__(self, list_id = 0):
        self._media_id = 0
        self._order, self._content = [], {}
        self.repeat = False

        self.list_id = list_id
        self.time_length = 0

    def __len__(self):
        return len(self._order)

    def _set_media_id(self):
        self._media_id += 1
        return self._media_id

    def _set_media_ans(self, id, pos = -1):
        ans = {"id": id, "pos": pos}
        try: ans.update(self._content[id])
        except KeyError:
            return None
        return ans

    def get(self, start = 0, stop = None):
        stop = stop or len(self._order)
        return map(self._set_media_ans, self._order[start:stop],\
                    range(start, stop))

    def get_item(self, id):
        try: return self._set_media_ans(id, self._order.index(id))
        except ValueError:
            return None

    def get_item_first(self):
        if self._order:
            return self._set_media_ans(self._order[0], 0)
        return None

    def get_item_last(self):
        if self._order:
            return self._set_media_ans(self._order[-1], len(self._order)-1)
        return None

    def get_ids(self):
        return self._order

    def reload_item_pos(self, media):
        try: pos = self._order.index(media["id"])
        except TypeError: return
        except ValueError:
            pos = -1
        media["pos"] = pos

    def set(self, medias):
        self._order, self._content = [], {}
        self.time_length = 0
        self.add_media(medias)

    def add_media(self, medias, first_pos = None):
        first_pos = first_pos or len(self._order)

        for index, m in enumerate(medias):
            id = self._set_media_id()
            self._order.insert(first_pos + index, id)
            self._content[id] = m
            # update medialist time length
            try: length = int(m["length"])
            except (ValueError, KeyError, TypeError):
                continue
            self.time_length += length
        self.list_id += 1

    def clear(self):
        self._order, self._content = [], {}
        self.time_length = 0
        self.list_id += 1

    def delete(self, ids, type = "id"):
        if type == "pos": ids = [self._order[p] for p in ids]
        missing_ids = [id for id in ids if id not in self._order]
        if missing_ids: return False

        for id in ids:
            self._order.remove(id)
            # update time length
            try: length = int(self._content[id]["length"])
            except (ValueError, KeyError, TypeError): continue
            else: self.time_length -= length
            del self._content[id]
        self.list_id += 1
        return True

    def next(self, media):
        try:
            idx = self._order.index(media["id"])
            id = self._order[idx+1]
        except (ValueError, KeyError):
            return None
        return self._set_media_ans(id, idx+1)

    def previous(self, media):
        try:
            idx = self._order.index(media["id"])
            id = idx and self._order[idx-1] or None
        except (ValueError, KeyError):
            return None
        return self._set_media_ans(id, idx-1)


class MediaList(SimpleMediaList):

    def move(self, ids, new_pos, type = "id"):
        if type == "pos": ids = [self._order[p] for p in ids]
        missing_ids = [id for id in ids if id not in self._order]
        if missing_ids: return False

        s_list = [id for id in self._order[:new_pos] if id not in ids]
        e_list = [id for id in self._order[new_pos:] if id not in ids]
        self._order = s_list + ids + e_list
        self.list_id += 1
        return True

    def shuffle(self, current = None):
        if not self._order: return
        random.shuffle(self._order)
        if current and current["id"]:
            try: self._order.remove(current["id"])
            except ValueError: pass
            else:
                self._order = [current["id"]] + self._order
        self.list_id += 1

    #
    # library changes action
    #
    def remove_media(self, media_id):
        ans = False
        for key, media in self._content.items():
            if media["media_id"] == media_id:
                self._order.remove(key)
                del self._content[key]
                ans = True
                try: length = int(media["length"])
                except (ValueError, KeyError, TypeError):
                    continue
                self.time_length -= length
        if ans: self.list_id += 1
        return ans

    def update_media(self, new_media):
        ans = False
        for key, m in self._content.items():
            if m["media_id"] == new_media["media_id"]:
                self._content[key] = new_media
                ans = True
        if ans: self.list_id += 1
        return ans


class ExtendedMediaList(MediaList):

    def __init__(self, list_id = 0):
        super(ExtendedMediaList, self).__init__(list_id)
        self.sorted = None
        self.filter_ = None
        self._filter = []

    def sort(self, key):
        def compare(a, b):
            return cmp(self._content[a][key], self._content[a][key])

        if self.sorted: self._order.reverse()
        try:
            self._order.sort(compare)
            self.sorted = key
        except KeyError:
            self.sorted = None
        self.list_id += 1

    def revert_sort(self):
        if self.sorted:
            self._order.reverse()
            self.list_id += 1

# vim: ts=4 sw=4 expandtab

File: webui--gwt/deejayd-webui/src/org/mroy31/deejayd/webui/medialist/MediaList.java
/*
 * Deejayd, a media player daemon
 * Copyright (C) 2007-2009 Mickael Royer <mickael.royer@gmail.com>
 *                         Alexandre Rossi <alexandre.rossi@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package org.mroy31.deejayd.webui.medialist;

import java.util.ArrayList;

import org.mroy31.deejayd.common.rpc.GenericRpcCallback;
import org.mroy31.deejayd.common.rpc.MediaFilter;
import org.mroy31.deejayd.webui.client.WebuiLayout;
import org.mroy31.deejayd.webui.resources.WebuiResources;

import com.google.gwt.core.client.GWT;
import com.google.gwt.event.dom.client.ScrollEvent;
import com.google.gwt.event.dom.client.ScrollHandler;
import com.google.gwt.event.shared.HandlerRegistration;
import com.google.gwt.json.client.JSONArray;
import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONValue;
import com.google.gwt.uibinder.client.UiBinder;
import com.google.gwt.uibinder.client.UiField;
import com.google.gwt.user.client.Command;
import com.google.gwt.user.client.DeferredCommand;
import com.google.gwt.user.client.IncrementalCommand;
import com.google.gwt.user.client.ui.CheckBox;
import com.google.gwt.user.client.ui.Composite;
import com.google.gwt.user.client.ui.FlexTable;
import com.google.gwt.user.client.ui.Image;
import com.google.gwt.user.client.ui.ScrollPanel;
import com.google.gwt.user.client.ui.Widget;

public class MediaList extends Composite {
    private final int PAGE_SIZE = 100;

    private static MediaListUiBinder uiBinder = GWT
            .create(MediaListUiBinder.class);
    interface MediaListUiBinder extends UiBinder<Widget, MediaList> {}

    private WebuiLayout ui;
    private String source;
    private boolean hasSelection = false;
    private MediaListRenderer renderer;
    private MediaFilter filter;
    private JSONArray mediaListArray;
    private int lastMediaLoaded = 0;
    private int currentPlaying = -1;

    @UiField FlexTable header;
    @UiField FlexTable mediaList;
    @UiField ScrollPanel mediaListPanel;
    @UiField(provided = true) final WebuiResources resources;

    /**
     * Rpc callback used for source.get command
     *
     */
    private class MediaListCallback extends GenericRpcCallback {
        public MediaListCallback(WebuiLayout ui) {
            super(ui);
        }
        public void onCorrectAnswer(JSONValue data) {
            // record filter
            JSONObject filterObj = data.isObject().get("filter").isObject();
            if (filterObj != null)
                filter = MediaFilter.parse(filterObj);
            else
                filter = null;
            JSONArray list = data.isObject().get("medias").isArray();
            setMedia(list);
        }
    }

    /**
     * Incremental command to load medialist
     *
     */
    private class MedialistUpdate implements IncrementalCommand {
        private final int CHUNK_SIZE = 500;
        private Command onFinish;
        private int lastItem;
        private int currentItem = lastMediaLoaded;

        public MedialistUpdate(int lastItem, Command cmd) {
            this.lastItem = lastItem;
            this.onFinish = cmd;
        }

        @Override
        public boolean execute() {
            currentItem = Math.min(lastItem, currentItem+CHUNK_SIZE);
            for (int idx=lastMediaLoaded; idx<currentItem; idx++) {
                JSONObject media = mediaListArray.get(idx).isObject();
                // set style for this row
                if ((idx % 2) == 0)
                    mediaList.getRowFormatter().setStyleName(idx,
                            resources.webuiCss().oddRow());

                renderer.formatRow(idx, mediaList, media);
            }

            lastMediaLoaded = currentItem;
            renderer.setLoadText(
                    ui.i18nMessages.itemLoadedDesc(lastMediaLoaded,
                    mediaListArray.size()));
            if (currentItem == lastItem) {
                mScrollReg = mediaListPanel.addScrollHandler(mScrollHandler);
                if (currentPlaying != -1 && currentPlaying < lastItem)
                    mediaList.getRowFormatter().addStyleName(currentPlaying,
                            resources.webuiCss().currentItem());

                if (onFinish != null)
                    onFinish.execute();
                return false;
            }
            return true;
        }
    }

    /**
     * Medialist scroll handler
     */
    private ScrollHandler mScrollHandler = new ScrollHandler() {

        @Override
        public void onScroll(ScrollEvent event) {
            int height = mediaList.getOffsetHeight();
            int scrollPosition = mediaListPanel.getScrollPosition();
            if ((height - scrollPosition) < 750) {
                loadMediaList();
            }
        }
    };
    private HandlerRegistration mScrollReg;

    /**
     * MediaList constructor
     * @param webui
     */
    public MediaList(WebuiLayout webui, String source) {
        this.source = source;
        this.ui= webui;
        this.resources = webui.resources;

        initWidget(uiBinder.createAndBindUi(this));
    }

    public void setOption(boolean hasSelection, MediaListRenderer r) {
        this.renderer = r;
        this.hasSelection = hasSelection;

        renderer.formatHeader(header, this);
        renderer.formatMediaList(mediaList);
    }

    public void update() {
        removeScrollHandler();

        mediaList.removeAllRows();
        mediaList.setWidget(0, 0, new Image(resources.loading()));
        mediaList.setText(0, 2, ui.i18nConstants.loading());
        ui.rpc.send(source+".get", new JSONArray(), new MediaListCallback(ui));
    }

    public void setMedia(JSONArray list) {
        renderer.setLoadText("");
        mediaList.removeAllRows();
        mediaListArray = list;
        lastMediaLoaded = 0;

        loadMediaList();
    }

    public void goTo(final int pos) {
        Command cmd =  new Command() {
            @Override
            public void execute() {
                Widget wg = mediaList.getWidget(pos, 0);
                mediaListPanel.ensureVisible(wg);
            }
        };

        if (pos < lastMediaLoaded)
            cmd.execute();
        else
            loadMediaList(pos+10, cmd);

    }

    public void checkRow(boolean value) {
        int size = mediaList.getRowCount();
        for (int idx=0; idx<size; idx++) {
            CheckBox ck = (CheckBox) mediaList.getWidget(idx, 0);
            ck.setValue(value);
        }
    }

    public MediaFilter getFilter() {
        return filter;
    }

    public String[] getSelection() {
        if (hasSelection) {
            ArrayList<String> selection = new ArrayList<String>();
            for (int idx=0; idx<mediaList.getRowCount(); idx++) {
                CheckBox ck = (CheckBox) mediaList.getWidget(idx, 0);
                if (ck.getValue())
                    selection.add(ck.getFormValue());
            }
            String[] result = new String[0];
            return selection.toArray(result);
        }
        return null;
    }

    public void setPlaying(int pos) {
        if (pos < lastMediaLoaded)
            mediaList.getRowFormatter().addStyleName(pos,
                    resources.webuiCss().currentItem());
        currentPlaying = pos;
    }

    public void resetPlaying(int pos) {
        if (pos < lastMediaLoaded)
            mediaList.getRowFormatter().removeStyleName(pos,
                    resources.webuiCss().currentItem());
        currentPlaying = -1;
    }

    private void loadMediaList() {
        loadMediaList(lastMediaLoaded+PAGE_SIZE);
    }

    private void loadMediaList(int lastItem) {
        loadMediaList(lastItem, null);
    }

    private void loadMediaList(int lastItem, Command onFinish) {
        removeScrollHandler();

        if (lastMediaLoaded < mediaListArray.size()) {
            renderer.setLoadText(ui.i18nConstants.loading());
            lastItem = Math.min(lastItem, mediaListArray.size());
            DeferredCommand.addCommand(new MedialistUpdate(lastItem, onFinish));

        }
    }

    private void removeScrollHandler() {
        if (mScrollReg != null) {
            mScrollReg.removeHandler();
            mScrollReg = null;
        }
    }
}

//vim: ts=4 sw=4 expandtab